README.md
# py-rcsb_ccmodels

Utilities for building structural models for RCSB Chemical Component definitions using
small molecule crystal structures in the CCDC and COD.

## Introduction

This package provides workflow utilities for generating search targets for RCSB
chemical component defintions, performing substructure searches on the CCDC database for
each search target, evaluating the these search results and building chemical component
structural models, and assembling selected models into a concatenated model dictionary.

This module depends on the RCSB CCDC wrapper module `py-rcsb_utils_ccdc` which in turn
depends on the proprietary CCDC Python API and CCDC/CSD database.
The latter dependencies require licenses and separate installation that is described
with the CCDC documentation.  The 2021 version of the CCDC API is supported only in Python 3.7
and the RCSB wrapper module provides a CLI that helps to isolate the current module
from this particular version requirement.

The Crystallographic Open Database (COD) provides a batch download of SMILES descriptors
that are seached against the Chemical Component Search database.  Metadata with experimental
descriptions and SDF files containing COD entry coordinates are downloaded for matching entries.
These data are used to build Chemical Component Model files that are integrated with model files
generated from matching CCDC entries.

### Installation

Download the library source software from the project repository and set the
enviroment corresponding appropriate to the CCDC and wrapper (`py-rcsb_utils_ccdc`)
in the provided script, `ccdc-api-env.sh`. This script provides an example for
a typical `pyenv` virtual configuration. To install from the source repository:

To install using pip:

```bash
# Install and configure the CCDC wrapper package
pip install rcsb.utils.ccdc
#
pip install  rcsb.ccmodels
```

Set the following environmental variables as required for you CCDC database
and Python API installation.

```bash
export CSDHOME=/Applications/CCDC/CSD_2021
# Locate your Python installation in which CCDC Python API is installed.
PYROOT=$PYENV_ROOT/versions/3.7.9
export CSD_PYTHON_ROOT_PATH=$PYROOT
#
# For Linux
export LD_LIBRARY_PATH=$PYROOT/lib:$PYROOT/lib/python3.7/site-packages/ccdc/_lib:$LD_LIBRARY_PATH
# For Macos
export DYLD_LIBRARY_PATH=$PYROOT/lib/python3.7/site-packages/ccdc/_lib
export DYLD_FRAMEWORK_PATH=$PYROOT/lib/python3.7/site-packages/ccdc/_lib
```

To install the packages from the source repository, follow these steps:

```bash
# Install the wrapper package
git clone --recurse-submodules https://github.com/rcsb/py-rcsb_utils_ccdc.git
cd py-rcsb_utils_ccdc
pip install -r requirements
pip install -e .
#

# Install the workflow utilities
git clone --recurse-submodules https://github.com/rcsb/py-rcsb_ccmodels.git
cd py-rcsb_ccmodels
pip install -r requirements
pip install -e .
# edit and set the enviroment in the following bash script ...
. ccdc-api-env.sh

```

Optionally, run test suite (currently Python versions 3.9) using
[setuptools](https://setuptools.readthedocs.io/en/latest/) or
[tox](http://tox.readthedocs.io/en/latest/example/platform.html):

```bash
  # edit and set the enviroment in the following bash script ...
  . ccdc-api-env.sh
  pip install -r requirements.txt
  python setup.py test

or simply run:

 # edit and set the enviroment in the following bash script ...
  . ccdc-api-env.sh
  tox
```

A CLI is provided to simplify access to the workflow steps.

```bash
# edit and set the enviroment in the following bash script ...
. ccdc-api-env.sh
#
python ChemCompModelExec.py --help
   -or-
cc_models_cli --help

usage: ChemCompModelExec.py [-h] [--generate_ccdc]
              [--cc_locator CC_LOCATOR] [--bird_locator BIRD_LOCATOR] [--prefix PREFIX]
              [--cache_path CACHE_PATH] [--use_cache] [--limit_perceptions] [--search_ccdc] [--update_only]
              [--build_ccdc] [--build_align_type BUILD_ALIGN_TYPE] [--search_cod] [--fetch_cod] [--build_cod]
              [--build_cod_timeout BUILD_COD_TIMEOUT] [--assemble] [--max_r_factor MAX_R_FACTOR]
              [--csdhome CSDHOME] [--python_lib_path PYTHON_LIB_PATH] [--python_version PYTHON_VERSION]
              [--num_proc NUM_PROC] [--chunk_size CHUNK_SIZE] [--search_timeout SEARCH_TIMEOUT] [--verbose]

optional arguments:
  -h, --help            show this help message and exit
  --generate_ccdc       Generate CCDC searchable files
  --cc_locator CC_LOCATOR
                        Chemical component reference dictioanary locator
  --bird_locator BIRD_LOCATOR
                        BIRD reference dictioanary locator
  --prefix PREFIX       Prefix for identifying reference resource files (e.g. abbrev)
  --cache_path CACHE_PATH
                        Top-level cache directory path
  --use_cache           Re-use cached resource files
  --limit_perceptions   Restrict automatic OE chemical perceptions
  --search_ccdc         Execute CCDC search
  --update_only         Only update current search results
  --build_ccdc          Build models from CCDC search results
  --build_align_type BUILD_ALIGN_TYPE
                        Alignment criteria (default: graph-relaxed-stereo-sdeq
  --search_cod          Execute COD search
  --fetch_cod           Fetch COD matching data
  --build_cod           Build models from COD search results
  --build_cod_timeout BUILD_COD_TIMEOUT
                        COD build time out (seconds) (Default None)
  --assemble            Assemble models into a concatenated file
  --max_r_factor MAX_R_FACTOR
                        Maximum permissible R-value in assembled model file (default=10.0)
  --csdhome CSDHOME     Path to the CSD release (path to CSD_202x)
  --python_lib_path PYTHON_LIB_PATH
                        Path to Python library
  --python_version PYTHON_VERSION
                        Python library version (default: 3.7)
  --num_proc NUM_PROC   Number of processes to execute (default=2)
  --chunk_size CHUNK_SIZE
                        Number of files loaded per process
  --search_timeout SEARCH_TIMEOUT
                        Search timeout (seconds) default=240
  --verbose             Verbose output
__________________________________________________
```

An example workflow script would look like the following:

```bash
#!/bin/bash
#
echo "Setup environment"
. ./ccdc-api-env.sh
#
echo "Begin search file generation"
cc_models_cli --generate --num_proc 6 --cache_path ./CACHE
#
echo "Begin search workflow"
cc_models_cli --search --num_proc 4 --cache_path ./CACHE
#
echo "Begin build workflow"
cc_models_cli --build --num_proc 4 --cache_path ./CACHE --build_align_type graph-relaxed-stereo-sdeq
#
echo "Begin COD search workflow"
cc_models_cli --search_cod --num_proc 12 --chunk_size 10 --cache_path ./CACHE
#
echo "Begin COD fetch data workflow"
cc_models_cli --fetch_cod --num_proc 12 --chunk_size 10 --cache_path ./CACHE
#
echo "Begin build workflow"
cc_models_cli --build_cod --num_proc 8 --chunk_size 20 --cache_path ./CACHE \
              --build_align_type graph-relaxed-stereo-sdeq --build_cod_timeout 120.0
#
echo "Begin assemble workflow"
cc_models_cli --assemble --min_r_factor 10.0  --cache_path ./CACHE
```


HISTORY.txt
# File: HISTORY.txt
#
26-Jan-2021 - V0.10 Preliminary version
27-Jan-2021 - V0.11 Update linux configuration and add exclusion filters in ChemCompModelGen()
27-Jan-2021 - V0.12 Python path adjustments in ChemCompModelExec()
27-Jan-2021 - V0.13 Make an index of the search files and avoid a file system glob search.
31-Jan-2021 - V0.14 Create models for protomeric matches with unmapped protons, multiprocess model build workflow
 1-Feb-2021 - V0.15 Add assemble workflow and model relabeling to permit sequential public model identifiers.
 2-Feb-2021 - V0.16 Adjust mismatch reporting during build workflow, update dependencies, add assemble check method
 2-Feb-2021 - V0.17 Add working path to OeAlignUtils(), adjust check method, keep heavy atom only structures
 3-Feb-2021 - V0.18 Simplify the build alignment criteria, add cli criteria option, and improve the handling of aromatic bond matching
 5-Feb-2021 - V0.19 Make assemble respect prior modelId assignments.
 5-Feb-2021 - V0.20 Add assemble workflow to cli in ChemCompModelExec()
 6-Feb-2021 - V0.21 Fix accession and audit reconciliation with prior model collections
 7-Feb-2021 - V0.22 Adjustments for handling BIRD style identifier and corrections to checking method.
 7-Feb-2021 - V0.23 Adjustments to identifier parsing method in ChemCompModelAssemble()
 9-Mar-2021 - V0.24 Add CODModelSearch and CODModelBuild modules
10-Mar-2021 - V0.25 Add multiprocessing methods for fetching COD data and entries
10-Mar-2021 - V0.26 Add COD functions to ChemCompModelExec().
10-Mar-2021 - V0.27 Split out COD fetch data from search in ChemCompModelExec().
10-Mar-2021 - V0.28 Add COD build results to Assemble task, add a timeout() on COD build alignment.
11-Mar-2021 - V0.29 Further additions of timeouts and data flow for COD build workflow.
12-Mar-2021 - V0.30 Add constraint on the maximum number of substructure matches in doAlignSS().
12-Mar-2021 - V0.31 Add constraint on the maximum number of substructure matches for OeDepictSubStructureAlignMultiPage().
13-Mar-2021 - V0.32 Refactor CODModelBuild(), change over to multiprocess based timeouts
14-Mar-2021 - V0.33 Adjustments for assembling COD and CCDC data
20-Jun-2023 - V0.34 Update PYROOT setting, timeOut setting, and env setup file, add example run file


rcsb/ccmodels/search/ChemCompModelExec.py
##
# File: ChemCompModelExec.py
# Date: 26-Jan-2021  jdw
#
#  Execution wrapper  --  for chemical component model construction workflow
#
#  Updates:
#
##
__docformat__ = "restructuredtext en"
__author__ = "John Westbrook"
__email__ = "jwest@rcsb.rutgers.edu"
__license__ = "Apache 2.0"

import argparse
import logging
import os
import sys

from rcsb.ccmodels.search import __version__
from rcsb.ccmodels.search.ChemCompModelAssemble import ChemCompModelAssemble
from rcsb.ccmodels.search.ChemCompModelBuild import ChemCompModelBuild
from rcsb.ccmodels.search.ChemCompModelGen import ChemCompModelGen
from rcsb.ccmodels.search.ChemCompModelSearch import ChemCompModelSearch
from rcsb.ccmodels.search.CODModelBuild import CODModelBuild
from rcsb.ccmodels.search.CODModelSearch import CODModelSearch

HERE = os.path.abspath(os.path.dirname(__file__))
TOPDIR = os.path.dirname(os.path.dirname(os.path.dirname(HERE)))

logging.basicConfig(level=logging.INFO, format="%(asctime)s [%(levelname)s]-%(module)s.%(funcName)s: %(message)s")
logger = logging.getLogger()
logger.setLevel(logging.INFO)


def main():
    parser = argparse.ArgumentParser()
    #
    parser.add_argument("--generate_ccdc", default=False, action="store_true", help="Generate CCDC searchable files")
    parser.add_argument("--cc_locator", default=None, help="Chemical component reference dictioanary locator")
    parser.add_argument("--bird_locator", default=None, help="BIRD reference dictioanary locator")
    parser.add_argument("--prefix", default=None, help="Prefix for identifying reference resource files (e.g. abbrev)")
    parser.add_argument("--cache_path", default=None, help="Top-level cache directory path")
    parser.add_argument("--use_cache", default=False, action="store_true", help="Re-use cached resource files")
    parser.add_argument("--limit_perceptions", default=False, action="store_true", help="Restrict automatic OE chemical perceptions")
    #
    parser.add_argument("--search_ccdc", default=False, action="store_true", help="Execute CCDC search")
    parser.add_argument("--update_only", default=False, action="store_true", help="Only update current search results")
    #
    parser.add_argument("--build_ccdc", default=False, action="store_true", help="Build models from CCDC search results")
    parser.add_argument("--build_align_type", default=False, action=None, help="Alignment criteria (default: graph-relaxed-stereo-sdeq")
    #
    parser.add_argument("--search_cod", default=False, action="store_true", help="Execute COD search")
    parser.add_argument("--fetch_cod", default=False, action="store_true", help="Fetch COD matching data")
    parser.add_argument("--build_cod", default=False, action="store_true", help="Build models from COD search results")
    parser.add_argument("--build_cod_timeout", default=None, help="COD build time out (seconds) (Default None)")
    #
    parser.add_argument("--assemble", default=False, action="store_true", help="Assemble models into a concatenated file")
    parser.add_argument("--max_r_factor", default=10.0, help="Maximum permissible R-value in assembled model file (default=10.0)")
    #
    parser.add_argument("--csdhome", default=None, help="Path to the CSD release (path to CSD_202x)")
    parser.add_argument("--python_lib_path", default=None, help="Path to Python library")
    parser.add_argument("--python_version", default=None, help="Python library version (default: 3.7)")
    #
    parser.add_argument("--num_proc", default=2, help="Number of processes to execute (default=2)")
    parser.add_argument("--chunk_size", default=10, help="Number of files loaded per process")
    parser.add_argument("--search_timeout", default=240, help="Search timeout (seconds) default=240")

    parser.add_argument("--verbose", default=False, action="store_true", help="Verbose output")

    args = parser.parse_args()
    #
    try:
        pyLib = args.python_lib_path if args.python_lib_path else os.path.join(os.environ["PYROOT"], "lib")
        pyRoot = os.path.dirname(pyLib)
        pyVer = args.python_version if args.python_version else "3.7"
        csdHome = args.csdhome if args.csdhome else os.environ["CSDHOME"]
        #
        doGenCcdc = args.generate_ccdc
        ccLocator = args.cc_locator
        birdLocator = args.bird_locator
        prefix = args.prefix
        cachePath = args.cache_path
        numProc = int(args.num_proc)
        chunkSize = int(args.chunk_size)
        useCache = args.use_cache
        limitPerceptions = args.limit_perceptions
        #
        doSearchCcdc = args.search_ccdc
        searchType = "substructure"
        updateOnly = args.update_only
        #
        doBuildCcdc = args.build_ccdc
        #
        doSearchCod = args.search_cod
        doFetchCod = args.fetch_cod
        doBuildCod = args.build_cod
        codBuildTimeOut = args.build_cod_timeout
        #
        doAssemble = args.assemble
        maxRFactor = args.max_r_factor
        verbose = args.verbose
        alignType = args.build_align_type if args.build_align_type else "graph-relaxed-stereo-sdeq"
        searchTimeOut = int(args.search_timeout)
    except Exception as e:
        logger.exception("Argument processing problem %s", str(e))
        parser.print_help(sys.stderr)
        exit(1)
    #
    try:
        os.environ["CSDHOME"] = csdHome
        os.environ["LD_LIBRARY_PATH"] = "%s:%s/python%s/site-packages/ccdc/_lib:$LD_LIBRARY_PATH" % (pyLib, pyLib, pyVer)
        os.environ["DYLD_LIBRARY_PATH"] = "%s/python%s/site-packages/ccdc/_lib" % (pyLib, pyVer)
        os.environ["DYLD_FRAMEWORK_PATH"] = "%s/python%s/site-packages/ccdc/_lib" % (pyLib, pyVer)

        logger.info("Using CSDHOME %s", os.environ["CSDHOME"])
        logger.info("Using DYLD_LIBRARY_PATH %s", os.environ["DYLD_LIBRARY_PATH"])
        logger.info("Using DYLD_FRAMEWORK_PATH %s", os.environ["DYLD_FRAMEWORK_PATH"])
        logger.info("Using source version %s", __version__)
        # -----------
        if doGenCcdc:
            ccmG = ChemCompModelGen(cachePath=cachePath, prefix=prefix)
            numMols = ccmG.buildSearchFiles(
                ccUrlTarget=ccLocator,
                birdUrlTarget=birdLocator,
                numProc=numProc,
                chunkSize=chunkSize,
                minCount=30,
                useCache=useCache,
                limitPerceptions=limitPerceptions,
            )
            logger.info("Generated %d search files using prefix %r", numMols, prefix)

        if doSearchCcdc:
            ccms = ChemCompModelSearch(cachePath=cachePath, pythonRootPath=pyRoot, csdHome=csdHome, prefix=prefix)
            rL = ccms.search(searchType, updateOnly=updateOnly, numProc=numProc, chunkSize=chunkSize, timeOut=searchTimeOut)
            logger.info("Search success count %d", len(rL))

        if doSearchCod:
            csU = CODModelSearch(cachePath=cachePath, numProc=numProc, useCache=True)
            csU.updateDescriptors()
            numMols = csU.search()
            logger.info("Search success count %d", numMols)

        if doFetchCod:
            csU = CODModelSearch(cachePath=cachePath, numProc=numProc, useCache=True)
            numMols = csU.fetchMatchedDataMp(useCache=True)
            logger.info("Fetch success count %d", numMols)

        if doBuildCcdc:
            ccmb = ChemCompModelBuild(cachePath=cachePath, prefix=prefix)
            rD = ccmb.build(alignType=alignType, numProc=numProc, chunkSize=chunkSize, verbose=verbose)
            logger.info("Built model count %d", len(rD))

        if doBuildCod:
            ccmb = CODModelBuild(cachePath=cachePath, prefix=prefix, timeOut=codBuildTimeOut)
            rD = ccmb.build(alignType=alignType, numProc=numProc, chunkSize=chunkSize)
            logger.info("Built model count %d", len(rD))

        if doAssemble:
            ccmb = ChemCompModelAssemble(cachePath=cachePath, prefix=prefix)
            numAssem = ccmb.assemble(maxRFactor=maxRFactor)
            logger.info("Assembled model count %d", numAssem)

    except Exception as e:
        logger.exception("Failing with %s", str(e))

    # ----------------------- - ----------------------- - ----------------------- - ----------------------- - ----------------------- -


if __name__ == "__main__":
    main()
